src/Symfony/Component/DependencyInjection/Compiler/AttributeAutoconfigurationPass.php

```diff
@@ -14,18 +14,65 @@
 use Symfony\Component\DependencyInjection\ChildDefinition;
 use Symfony\Component\DependencyInjection\ContainerBuilder;
 use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Exception\LogicException;
 
 /**
  * @author Alexander M. Turek <me@derrabus.de>
  */
 final class AttributeAutoconfigurationPass extends AbstractRecursivePass
 {
+    private $classAttributeConfigurators = [];
+    private $methodAttributeConfigurators = [];
+    private $propertyAttributeConfigurators = [];
+    private $parameterAttributeConfigurators = [];
+
     public function process(ContainerBuilder $container): void
     {
         if (80000 > \PHP_VERSION_ID || !$container->getAutoconfiguredAttributes()) {
             return;
         }
 
+        foreach ($container->getAutoconfiguredAttributes() as $attributeName => $callable) {
+            $callableReflector = new \ReflectionFunction(\Closure::fromCallable($callable));
+            if ($callableReflector->getNumberOfParameters() <= 2) {
+                $this->classAttributeConfigurators[$attributeName] = $callable;
+                continue;
+            }
+
+            $reflectorParameter = $callableReflector->getParameters()[2];
+            $parameterType = $reflectorParameter->getType();
+            $types = [];
+            if ($parameterType instanceof \ReflectionUnionType) {
+                foreach ($parameterType->getTypes() as $type) {
+                    $types[] = $type->getName();
+                }
+            } elseif ($parameterType instanceof \ReflectionNamedType) {
+                $types[] = $parameterType->getName();
+            } else {
+                throw new LogicException(sprintf('Parameter "$%s" in callable passed to registerAttributeForAutoconfiguration() in "%s" on line "%d" should have a type. Use one or more of the following types: \ReflectionClass|\ReflectionMethod|\ReflectionProperty|\ReflectionParameter.', $reflectorParameter->getName(), $callableReflector->getFileName(), $callableReflector->getStartLine()));
+            }
+
+            if (\in_array(\Reflector::class, $types, true)) {
```

Comment:
thanks for adding this

I'm actively thinking about @Tobion's feedback and I think we can accommodate both our approaches by filtering based on both the attributes' targets and the allowed tyep hints.

Note that I don't know how to read the `#[Attribute]` attribute on attribute classes...

Can you figure out how? Anyoneelse?

Reply:
This should do it:
```php
try {
    $r = new \ReflectionClass($attributeName);
} catch (\ReflectionException $e) {
    continue;
}

$targets = $r->getAttributes(\Attribute::class)[0] ?? 0;
$targets = $targets ? $targets->getArguments()[0] ?? -1 : 0;
```

Reply:
we could even throw when $targets and the type hints of an argument have an empty intersection

Reply:
I think it's too complicate to support both.

My suggestion would be:
* Read targets from Attribute with the above snippet, and use that
* Typehint in phpdoc is `\Reflector` but can also be made specific (\ReflectionMethod e.g.) or a union type. It's not read by Symfony. 

@nicolas-grekas @Tobion If you agree with this I will update the PR.

Reply:
I'm AFK for two weeks and I'd like to see how supporting both would work. I can give it a try when I'll be back!

Reply:
@nicolas-grekas If you can give it a try, please do. Thanks 🙏

Reply:
@nicolas-grekas Hope you had a good time hiking in Greece. Curious if you already have an idea how to solve this?